### PR TITLE
BSA: fuse single-group SM100 blk64 dK/dV epilogue

### LINKED ISSUE #1089: Optimize SM100 blk64 BSA backward single-group dK/dV epilogue
## Problem

The SM100/SM103 blk64 BSA backward path always accumulates dK and dV through FP32 global workspaces and atomic adds, then launches separate conversion work, even when the bucketed reverse CSR has exactly one Q group. In that case each `(batch, head, KV64)` tile has one unique CTA owner, so cross-group dK/dV accumulation is unnecessary.

## Proposed optimization

Mirror the existing SM100 blk128 policy in the blk64 path:

- specialize at compile time when `num_q_groups == 1`;
- keep dK/dV in FP32 TMEM across the complete reverse row;
- scale dK, convert dK/dV to BF16, and write caller outputs directly;
- retain the current FP32 atomic workspace and postprocess path for multiple Q groups.

This does not require a public API change. Empty KV rows, partial block handling, BHSD/BSHD layouts, and the generic multi-group path must retain current behavior.

## Prior evidence

An independent SM103 Q64/KV64 implementation measured the isolated direct-epilogue change at approximately 4.5% lower end-to-end latency, with two 128 MiB FP32 workspaces removed at `B=1,H=32,S=8192,D=128` and about 31.6% fewer measured DRAM writes. These numbers are directional; the change will be benchmarked directly in this repository before making an upstream performance claim.

I plan to submit a focused PR with single- versus multi-group correctness coverage and short/sustained B30Z measurements.

### PR DESCRIPTION
**Pull request by wplf**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` on every changed file.
- [x] I reviewed the applicable Hard Rules in `AGENTS.md`.
- [ ] I added the required labels; external contributors may need a maintainer to apply them.

## Affected area

FE OSS kernels or CuTeDSL — SM100/SM103 blk64 block-sparse attention backward.

## Summary

Specialize the blk64 backward epilogue when the bucketed reverse CSR has exactly one Q group and the physical K extent is KV64-aligned:

- keep the uniquely owned dK/dV accumulators in FP32 TMEM across the complete reverse row;
- scale dK, convert dK/dV to BF16, and write caller outputs directly from the owner CTA;
- remove both FP32 dK/dV workspaces from that specialization;
- use the existing dQ conversion kernel to zero empty KV rows;
- retain the existing FP32 atomic workspace/postprocess path for multiple Q groups and sequence tails.

The compile key includes the epilogue policy, so direct and generic artifacts cannot alias. The public API and default numerical contract are unchanged.

This also restores the `torch` import used by `_device_capability()` in `bucketed_k2q_csr.py`; current `develop` otherwise raises `NameError` while building reverse CSR metadata after #1073.

## Why

With one Q group, every `(batch, head, KV64)` tile has exactly one CTA owner. Atomic FP32 dK/dV accumulation and two full-size workspaces are unnecessary in that case. The blk128 implementation already uses this ownership optimization; this change applies the same policy to blk64 while preserving the generic path where several Q groups contribute to one KV tile.

## Performance

NVIDIA B30Z / SM103, BF16 `B=1,H=32,Sq=Sk=8192,D=128`, exact 50% K64 sparsity, one prebuilt reverse-CSR group, BHSD, preallocated outputs. Each result is an ABBA midpoint from isolated processes based on `develop@fa7a433e`. Backward throughput uses the active five-GEMM `10D` numerator.

```text
20 warmups / 500 samples
  baseline:  2.147920 ms / 639.87 TFLOP/s
  candidate: 2.009728 ms / 683.87 TFLOP/s
  change:    -6.43% latency / +6.88% throughput

100 warmups / 5,000 samples
  baseline:  2.148560 ms / 639.68 TFLOP/s
  candidate: 2.043856 ms / 672.45 TFLOP/s
  change:    -4.87% latency / +5.12% throughput

peak live temporary allocation
  baseline:  386 MiB
  candidate: 130 MiB
  change:   -256 MiB
```

Forced two-Q-group control (`bucket_size_blocks=64`, 20x500):

```text
baseline:  2.248416 ms
candidate: 2.248864 ms
```

The generic path is unchanged within measurement noise.

## Related issues

Fixes #1089.

## API and compatibility impact

None. The direct path is selected internally only for a single reverse-CSR Q group with an aligned physical K extent. Multi-group and K-tail calls retain the existing implementation.

## Testing

Environment: B30Z/SM103, PyTorch 2.13.0+cu130, CuTe DSL 4.6.2, cuDNN backend 9.20.

```bash
cd test/python
pytest -q -m "L0 or L1 or L2" \
  fe_api/bsa/test_BSA_attention_forward.py \
  fe_api/bsa/test_BSA_attention_backward.py
# 24 passed, 23 skipped
```

The expanded blk64 test covers:

- direct single-group and forced multi-group dispatch;
- dirty preallocated outputs and empty KV owners;
- partial `block_sizes`;
- custom softmax scale;
- non-KV64-aligned sequence-tail fallback;
- numerical comparison against the FP32 reference.

The dispatch assertion was verified RED on unmodified `develop` (`[True] != [False]`) and green on this branch. A manual BSHD direct-versus-generic comparison measured maximum differences of 0 for dQ, `2.98e-8` for dK, and `3.05e-5` for dV.

`pre-commit` passed for all five changed files. SM100/B200, SM110, multi-GPU, and JAX were not tested locally.

## Summary by CodeRabbit

- **Performance**
  - Improved block-sparse attention backward performance for aligned, single-group workloads by writing dK/dV gradients directly when possible.
  - Retained accumulation fallback behavior for multi-group and uneven sequence layouts.

- **Bug Fixes**
  - Improved handling of empty KV blocks, including proper zero initialization of gradient outputs.

- **Tests**
  - Expanded coverage for direct and fallback gradient paths, custom softmax scales, sequence tails, empty blocks, and in-place gradient outputs.

### CODE REVIEW

**Comment by wplf** [comment]
As an external contributor I cannot set repository labels, Milestone, or Projects. Could a maintainer please apply `cat-perf-bug`, `area:sparse_attention`, and `orig-external`, assign the appropriate Milestone/Project, and run the OSS CI after review? Local SM103 validation and the full performance protocol are included in the PR description.

**Comment by YangXu1990uiuc** [comment]
I am starting OSS CI for `af55c52c5fb5` and approving the pending fork Style check. The following standalone `@cudnn-ci-bot run oss` comment is a Codex bot CI action.

For future updates, request a maintainer to post that exact standalone command if you lack trigger permission; fork GitHub Actions may also need a maintainer's approval. This revision inherits the develop SDPA stats enum build issue, so build failures need to be distinguished from the OSS GPU test results.

Codex bot review — model `gpt-6-astra`.

**Comment by YangXu1990uiuc** [comment]
I am requesting test CI for `af55c52c5fb5` with the following standalone command: `@cudnn-ci-bot run oss`. For future updates, request the appropriate test target before review; a maintainer may need to trigger it.

Codex bot review — model `gpt-6-astra`.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `af55c52c5fb5` against base `2cf27d377e06`.

The single-owner specialization is sound: the direct path is restricted to one Q group and aligned KV tiles, the compile key distinguishes the policy, and empty KV owners are explicitly zeroed. I found no defect introduced by this change that blocks approval.

Validation: Independently ran the BSA forward/backward selection on SM100: 24 passed and 23 architecture-specific skips with CuTe DSL 4.8; the same 24/23 result at the supported DSL 4.6.2 floor. Additional DSL 4.6.2 probes passed for B=2/H=3, BHSD and physically contiguous BSHD, varying sparse counts, partial blocks, poisoned/guarded outputs and exact-zero empty KV rows. A 17-Q-block reverse row also passed. ABBA SM100 spot check, BF16 B=1/H=8/Sq=Sk=2048/D=128, 50% sparsity, prebuilt reverse CSR and preallocated outputs: single-group GPU replay fell from 175.33-175.38 us to 147.51-147.55 us; workspace fell from 25,296,896 to 8,519,680 bytes. Two-group control stayed around 195 us. CPU enqueue medians overlapped (single group 75.5-78.0 us baseline, 76.7-78.7 us head). Correctness and changed-input/poisoned-output CUDA Graph replay passed before timing. Baseline needed only the missing torch import supplied to execute its old kernel; these are backward-component measurements, not end-to-end model results. Started OSS pipeline 68138680 and approved the fork Style check. Style, all builds, all three C++ sanitizer lanes, OSS SM80/SM90, SM100 shard1 and Rubin passed. SM100 shard0 has only the two existing JAX BSA/DSL4.6.2 import failures (2,062 passed); API-index analysis also remains red. The CI mirror includes newer develop fixes, so this is integration evidence alongside the independent head-only tests above.

Limitations: A separate Sq=131 Q-tail probe fails in both develop and this head on identical inputs, with identical maximum dQ/dK/dV errors (0.4371/0.1407/0.1876). This predates the epilogue change; the passing K-tail regression does not establish arbitrary Q-tail correctness. Independent local hardware coverage is SM100. The independent tests do not establish SM103/SM110 coverage; JAX coverage remains limited by the CI dependency-floor failures.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.
